src/GC/GC.jl

```diff
@@ -9,77 +9,116 @@ module GC
 """
 function disable()
     ENABLED[] = false
+    reset(ENABLED_EVENT)
     return
 end
 
 """
     PythonCall.GC.enable()
 
-Re-enable the PythonCall garbage collector.
-
-This frees any Python objects which were finalized while the GC was disabled, and allows
-objects finalized in the future to be freed immediately.
+Re-enable the PythonCall garbage collector. This should generally not be required.
 
-Like most PythonCall functions, you must only call this from the main thread.
 """
 function enable()
     ENABLED[] = true
+    notify(ENABLED_EVENT)
+    return
+end
+
+function enqueue(ptr::C.PyPtr)
+    if ptr != C.PyNULL && C.CTX.is_initialized    
+        put!(QUEUE, ptr)
+    end
+    return
+end
+
+function enqueue_all(ptrs)
+    if C.CTX.is_initialized
+        for ptr in ptrs
+            put!(QUEUE, ptr)
+        end
+    end
+    return
+end
+
+# must only be called from thread 1 by the task in `GC_TASK[]`
+function unsafe_process_queue!()
     if !isempty(QUEUE)
         C.with_gil(false) do
-            for ptr in QUEUE
+            while !isempty(QUEUE) && ENABLED[]
+                # This should never block, since there should
+                # only be one consumer
+                # (we would like to not block while holding the GIL)
+                ptr = take!(QUEUE)
                 if ptr != C.PyNULL
                     C.Py_DecRef(ptr)
                 end
             end
         end
     end
-    empty!(QUEUE)
-    return
+    return nothing
 end
 
-function enqueue(ptr::C.PyPtr)
-    if ptr != C.PyNULL && C.CTX.is_initialized
-        if ENABLED[]
-            C.with_gil(false) do
-                C.Py_DecRef(ptr)
-            end
-        else
-            push!(QUEUE, ptr)
+function gc_loop()
+    while true
+        if ENABLED[] && !isempty(QUEUE)
+            unsafe_process_queue!()
+            # just for testing purposes
+            Base.@lock GC_FINISHED notify(GC_FINISHED)
         end
+        # wait until there is both something to process
+        # and GC is `enabled`
+        wait(QUEUE)
+        wait(ENABLED_EVENT)
     end
-    return
 end
 
-function enqueue_all(ptrs)
-    if C.CTX.is_initialized
-        if ENABLED[]
-            C.with_gil(false) do
-                for ptr in ptrs
-                    if ptr != C.PyNULL
-                        C.Py_DecRef(ptr)
-                    end
-                end
-            end
-        else
-            append!(QUEUE, ptrs)
-        end
+function launch_gc_task()
+    if isassigned(GC_TASK) && Base.istaskstarted(GC_TASK[]) && !Base.istaskdone(GC_TASK[])
+        throw(ConcurrencyViolationError("PythonCall GC task already running!"))
     end
-    return
+    task = Task(gc_loop)
+    task.sticky = VERSION >= v"1.7" # disallow task migration which was introduced in 1.7
+    # ensure the task runs from thread 1
+    ccall(:jl_set_task_tid, Cvoid, (Any, Cint), task, 0)
+    schedule(task)
```

Comment:
this is from https://github.com/tro3/ThreadPools.jl/blob/44730372b9b00ba38192e5d2850dfe24c20cdbeb/src/macros.jl#L269-L276